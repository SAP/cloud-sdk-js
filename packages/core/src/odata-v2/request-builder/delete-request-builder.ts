import {
  Constructable,
  DeleteRequestBuilder as DeleteRequestBuilderBase,
  FieldType
} from '../../odata-common';
import { Entity } from '../entity';
import { oDataUri } from '../uri-conversion';
/**
 * Create OData query to delete an entity.
 *
 * @typeparam EntityT - Type of the entity to be deleted
 */
export class DeleteRequestBuilder<
  EntityT extends Entity
> extends DeleteRequestBuilderBase<EntityT> {
  /**
   * Creates an instance of DeleteRequestBuilder. If the entity is passed, version identifier will also be added.
   *
   * @param entityConstructor - Constructor type of the entity to be deleted
   * @param keysOrEntity - Entity or Key-value pairs of key properties for the given entity
   */
  constructor(
    entityConstructor: Constructable<EntityT>,
    keysOrEntity: Record<string, FieldType> | Entity
  ) {
    super(entityConstructor, oDataUri, keysOrEntity);
  }

  /**
   * Add an eTag version identifier in the delete request header.
   *
   * @param eTag - The version identifier of the entity
   * @returns The builder itself, to facilitate method chaining
   */
  setVersionIdentifier(eTag: string): this {
    this.requestConfig.eTag = eTag;
    return this;
  }
}

export { DeleteRequestBuilder as DeleteRequestBuilderV2 };
