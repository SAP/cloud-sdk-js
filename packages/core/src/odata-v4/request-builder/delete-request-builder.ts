import {
  Constructable,
  DeleteRequestBuilder as DeleteRequestBuilderBase,
  EntityIdentifiable,
  FieldType
} from '../../odata-common';
import { Entity } from '../entity';
import { oDataUri } from '../uri-conversion';
/**
 * Create OData query to delete an entity.
 *
 * @typeparam EntityT - Type of the entity to be deleted
 */
export class DeleteRequestBuilder<EntityT extends Entity>
  extends DeleteRequestBuilderBase<EntityT>
  implements EntityIdentifiable<EntityT> {
  readonly _entityConstructor: Constructable<EntityT>;
  readonly _entity: EntityT;

  /**
   * Creates an instance of DeleteRequestBuilder. If the entity is passed, version identifier will also be added.
   *
   * @param entityConstructor - Constructor type of the entity to be deleted
   * @param keysOrEntity - Entity or Key-value pairs of key properties for the given entity
   */
  constructor(
    entityConstructor: Constructable<EntityT>,
    keysOrEntity: Record<string, FieldType> | Entity
  ) {
    super(entityConstructor, oDataUri, keysOrEntity);
  }

  /**
   * Add ETag version identifier in the delete request header.
   *
   * @param etag - The version identifier of the entity
   * @returns The builder itself, to facilitate method chaining
   */
  setVersionIdentifier(etag: string): this {
    if (etag) {
      this.requestConfig.addCustomHeaders({ 'if-match': etag });
    }
    return this;
  }
}

export { DeleteRequestBuilder as DeleteRequestBuilderV4 };
