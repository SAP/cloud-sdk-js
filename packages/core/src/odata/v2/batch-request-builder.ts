/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import { errorWithCause, MapType } from '@sap-cloud-sdk/util';
import { v4 as uuid } from 'uuid';
import {
  Destination,
  DestinationNameAndJwt,
  DestinationOptions,
  toDestinationNameUrl
} from '../../scp-cf';
import { Constructable, BatchResponse } from '../common';
import { MethodRequestBuilderBase } from '../common/request-builder/request-builder-base';
import { ODataBatchRequestConfig } from '../common/request/odata-batch-request-config';
import { EntityV2 } from './entity';
import {
  ODataBatchChangeSetV2,
  serializeBatchRequest
} from './batch-request-serializer';
import {
  CreateRequestBuilderV2,
  DeleteRequestBuilderV2,
  GetAllRequestBuilderV2,
  GetByKeyRequestBuilderV2,
  UpdateRequestBuilderV2
} from './request-builder';
import {
  detectNewLineSymbol,
  partitionBatchResponse,
  buildRetrieveOrErrorResponse,
  buildWriteResponses
} from './batch-response-parser';

/**
 * Create a batch request to invoke multiple requests as a batch. The batch request builder accepts retrieve requests, i. e. [[GetAllRequestBuilder | getAll]] and [[GetByKeyRequestBuilder | getByKey]] requests and change sets, which in turn can contain [[CreateRequestBuilder | create]], [[UpdateRequestBuilder | update]] or [[DeleteRequestBuilder | delete]] requests.
 * The retrieve and change sets will be excuted in order, while the order within a change set can vary.
 */
export class ODataBatchRequestBuilderV2 extends MethodRequestBuilderBase<
  ODataBatchRequestConfig
> {
  /**
   * Creates an instance of ODataBatchRequestBuilder.
   *
   * @param defaultServicePath - Service path
   * @param requests - An array of retrieve requests or change sets
   * @param entityToConstructorMap - A map that holds the entity type to constructor mapping
   */
  constructor(
    readonly defaultServicePath: string,
    readonly requests: (
      | ODataBatchChangeSetV2<
          | CreateRequestBuilderV2<EntityV2>
          | UpdateRequestBuilderV2<EntityV2>
          | DeleteRequestBuilderV2<EntityV2>
        >
      | GetAllRequestBuilderV2<EntityV2>
      | GetByKeyRequestBuilderV2<EntityV2>
    )[],
    readonly entityToConstructorMap: MapType<Constructable<EntityV2>>
  ) {
    super(new ODataBatchRequestConfig(defaultServicePath, uuid()));
    this.requestConfig.payload = serializeBatchRequest(this);
  }

  /**
   * Execute the given request and return the according promise. Please notice: The sub-requests may fail even the main request is successful.
   *
   * @param destination - Targeted destination on which the request is performed.
   * @param options - Options to employ when fetching destinations.
   * @returns Promise resolving to the requested data.
   */
  async execute(
    destination: Destination | DestinationNameAndJwt,
    options?: DestinationOptions
  ): Promise<BatchResponse[]> {
    return this.build(destination, options)
      .then(request => request.execute())
      .then(response =>
        buildResponses(
          partitionBatchResponse(response),
          this.entityToConstructorMap,
          detectNewLineSymbol(response.data)
        )
      )
      .catch(error =>
        Promise.reject(
          errorWithCause(
            `The batch request failed! The destination provided: ${toDestinationNameUrl(
              destination
            )} and the options: ${options}.`,
            error
          )
        )
      );
  }
}

function buildResponses(
  responses: string[],
  entityToConstructorMap: MapType<Constructable<EntityV2>>,
  lineBreak: string
): BatchResponse[] {
  return responses.map(r =>
    buildResponse(r, entityToConstructorMap, lineBreak)
  );
}

function buildResponse(
  response: string,
  entityToConstructorMap: MapType<Constructable<EntityV2>>,
  lineBreak: string
) {
  if (isChangeSet(response)) {
    return buildWriteResponses(response, entityToConstructorMap, lineBreak);
  }
  if (isRetrieveRequestOrError(response)) {
    return buildRetrieveOrErrorResponse(
      response,
      entityToConstructorMap,
      lineBreak
    );
  }
  throw Error(
    `The response: ${JSON.stringify(
      response
    )} is not a valid retrieve request or change set, because it does not contain the proper Content-Type.`
  );
}

function isChangeSet(response: string): boolean {
  return !!response.match(/Content-Type: multipart\/mixed;/);
}

function isRetrieveRequestOrError(response: string): boolean {
  return !!response.match(/Content-Type: application\/http/);
}

export { ODataBatchRequestBuilderV2 as ODataBatchRequestBuilder };
