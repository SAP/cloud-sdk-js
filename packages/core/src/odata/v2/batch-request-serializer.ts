/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import {
  serializeChangeSet,
  serializeRequest
} from '../common/request-builder/batch/batch-request-serializer';
import { BatchChangeSet } from '../common/request-builder/batch/batch-change-set';
import { EntityV2 } from './entity';
import {
  CreateRequestBuilderV2,
  DeleteRequestBuilderV2,
  GetAllRequestBuilderV2,
  GetByKeyRequestBuilderV2,
  UpdateRequestBuilderV2
} from './request-builder';

/**
 * @deprecated Since v1.30.0. Use [[serializeRequest]] instead.
 * Build a string as the request body of the retrieve request.
 * Below is an example of the generated body, where the two empty line are mandatory to make the request valid.
 * *** example starts ***
 * Content-Type: application/http
 * Content-Transfer-Encoding: binary
 *
 * GET /SomeUrl/API_BUSINESS_PARTNER/A_BusinessPartnerBank?$format=json&$top=1 HTTP/1.1
 *
 *
 * *** example ends ***
 * @param request - The request builder of the retrieve request.
 * @returns The request body.
 */
export function toBatchRetrieveBodyV2(
  request: GetAllRequestBuilderV2<EntityV2> | GetByKeyRequestBuilderV2<EntityV2>
): string {
  return serializeRequest(request);
}

/**
 * @deprecated Since v1.30.0. Use [[serializeChangeSet]] instead.
 * Serialize change set to string.
 * @param changeSet - Change set containing a collection of write operations.
 * @returns The serialized string representation of a change set.
 */
export function toBatchChangeSetV2<
  T extends
    | CreateRequestBuilderV2<EntityV2>
    | UpdateRequestBuilderV2<EntityV2>
    | DeleteRequestBuilderV2<EntityV2>
>(changeSet: BatchChangeSet<T>): string | undefined {
  return serializeChangeSet(changeSet);
}

export { toBatchChangeSetV2 as toBatchChangeSet };
export { toBatchRetrieveBodyV2 as toBatchRetrieveBody };
