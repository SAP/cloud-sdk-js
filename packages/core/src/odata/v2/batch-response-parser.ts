/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
import { last, MapType, createLogger } from '@sap-cloud-sdk/util';
import { HttpResponse } from '../../http-client';
import { getHeaderValue } from '../../header-builder';
import {
  getSingleResult,
  isCollectionResult,
  getCollectionResult
} from '../v2/request-builder/response-data-accessor';
import {
  Constructable,
  ErrorResponse,
  ReadResponse,
  WriteResponses,
  WriteResponse
} from '../common';
import { EntityV2 } from './entity';
import { deserializeEntityV2 } from './entity-deserializer';

const logger = createLogger({
  package: 'core',
  messageContext: 'batch-response-parser'
});

/**
 * Detects the system dependent line break in a string.
 * @param str The string to check for line breaks. Should have at least two lines, otherwise an error will be thrown.
 * @returns The system dependent line break
 */
export function detectNewLineSymbol(str: string): string {
  if (str.includes('\r\n')) {
    return '\r\n';
  }
  if (str.includes('\n')) {
    return '\n';
  }
  throw new Error('Cannot detect line breaks in the batch response body.');
}

/**
 * Get the response body from the string representation of a response.
 * @param response String representation of a response.
 * @returns The response body as a one line string.
 */
export function getResponseBody(response: string): string {
  const newLineSymbol = detectNewLineSymbol(response);
  const lines = response.split(newLineSymbol);

  // A valid response should contain at least three lines, part id, empty line and response body.
  if (lines.length >= 3) {
    return lines[lines.length - 1];
  }

  throw Error(
    `Cannot parse batch subrequest response body. Expected at least three lines in the response, got ${lines.length}.`
  );
}

/**
 * Parse the string representation of response headers into an object
 * @param headersStr Header string representation
 * @returns The headers as an object.
 */
function parseHeaders(headersStr: string): Record<string, any> {
  const newLineSymbol = detectNewLineSymbol(headersStr);
  return headersStr.split(newLineSymbol).reduce((headers, line) => {
    const [key, value] = line.split(':');
    return { ...headers, [key]: value };
  }, {});
}

/**
 * Get the boundary from the content type header value.
 * @param contentType Value of the content type header
 * @returns The boundary or undefined if none is given.
 */
function getBoundary(contentType: string | undefined): string | undefined {
  if (contentType) {
    return last(contentType.split('boundary='));
  }
}

/**
 * Split a batch response into an array of sub responses for the retrieve requests and changesets.
 * @param response The raw HTTP response.
 * @returns A list of sub responses represented as strings.
 */
export function splitBatchResponse(response: HttpResponse): string[] {
  const body = response.data.trim();
  if (!body) {
    return [];
  }

  const boundary = getBoundary(
    getHeaderValue('content-type', response.headers)
  );

  if (!boundary) {
    throw new Error(
      'Could not parse batch response. Expected multipart response with boundary.'
    );
  }

  return splitResponse(body, boundary);
}

/**
 * Split a changeset (sub) response into an array of sub responses.
 * @param changeSetResponse The string representation of a change set response.
 * @returns A list of sub responses represented as strings.
 */
export function splitChangeSetResponse(changeSetResponse: string): string[] {
  const [changeSetResponseHeader] = changeSetResponse.split('--');

  const headers = parseHeaders(changeSetResponseHeader);
  const boundary = getBoundary(getHeaderValue('content-type', headers));

  if (!boundary) {
    throw Error('Cannot parse change set.');
  }

  return splitResponse(changeSetResponse, boundary);
}

/**
 * Split a string representation of a response into sub responses given its boundary.
 * @param response The string representation of the response to split.
 * @param boundary The boundary to split by.
 * @returns A list of sub responses represented as strings.
 */
export function splitResponse(response: string, boundary: string): string[] {
  const parts = response.split(`--${boundary}`).map(part => part.trim());

  if (parts.length >= 3) {
    return parts.slice(1, parts.length - 1);
  }

  throw new Error(
    'Could not parse batch response body. Expected at least two response boundaries.'
  );
}

/**
 * Parse the entity name from the metadata uri. This should be the `__metadata` property of a single entity in the response.
 * @param uri The URI to parse the entity name from
 * @returns The entity name.
 */
export function parseEntityNameFromMetadataUri(
  uri: string | undefined
): string {
  if (!uri) {
    throw new Error(
      `Could not retrieve entity name from metadata. URI was: '${uri}'.`
    );
  }
  const [pathBeforeQuery] = uri.split('?');
  const [pathBeforeKeys] = pathBeforeQuery.split('(');
  const uriParts = pathBeforeKeys.split('/');

  // Remove another part in case of a trailing slash
  const entityName = uriParts.pop() || uriParts.pop();
  if (!entityName) {
    throw Error(
      `Could not retrieve entity name from metadata. Unknown URI format. URI was: '${uri}'.`
    );
  }
  return entityName;
}

// version specific
/**
 * Retrieve the constructor for a specific single response body.
 * @param responseBody The body of a single response as an object.
 * @param entityToConstructorMap Mapping between entity names and their respective constructors.
 * @returns The constructor if found in the mapping, undefined otherwise.
 */
export function getConstructor(
  responseBody: Record<string, any>,
  entityToConstructorMap: MapType<Constructable<EntityV2>>
): Constructable<EntityV2> | undefined {
  const entityJson = isCollectionResult(responseBody)
    ? getCollectionResult(responseBody)[0]
    : getSingleResult(responseBody);

  const entityUri = entityJson?.__metadata?.uri;
  if (entityUri) {
    return entityToConstructorMap[parseEntityNameFromMetadataUri(entityUri)];
  }

  logger.warn('Could not parse constructor from response body.');
}

/**
 * Parse the HTTP code of response.
 * @param response String representation of the response.
 * @returns The HTTP code.
 */
export function parseHttpCode(response: string): number {
  const group = response.match(/HTTP\/\d\.\d (\d{3}).*?/);

  if (group) {
    return parseInt(group[1].toString());
  }

  throw new Error('Cannot parse http code of the response.');
}

/**
 * Create a function to transform the parsed response body to a list of entities of the given type or an error.
 * @param body The parsed JSON reponse body.
 * @returns A function to be used for transformation of the read response.
 */
function asReadResponse(body: any) {
  return <T extends EntityV2>(constructor: Constructable<T>): Error | T[] => {
    if (body.error) {
      return new Error(body.error);
    }
    if (isCollectionResult(body)) {
      return getCollectionResult(body).map(r =>
        deserializeEntityV2(r, constructor)
      );
    }
    return [deserializeEntityV2(getSingleResult(body), constructor)];
  };
}

/**
 * Create a function to transform the parsed response body to an entity of the given type.
 * @param body The parsed JSON reponse body.
 * @returns A function to be used for transformation of the write response.
 */
function asWriteResponse(body) {
  return <T extends EntityV2>(constructor: Constructable<T>) =>
    deserializeEntityV2(getSingleResult(body), constructor);
}

/**
 * Get the body from the given response and parse it to JSON.
 * @param response The string representation of a single response.
 * @returns The parsed JSON representation of the response body.
 */
function parseResponseBody(response: string): Record<string, any> {
  const responseBody = getResponseBody(response);
  if (responseBody) {
    try {
      return JSON.parse(responseBody);
    } catch (err) {
      logger.error(
        `Could not parse response body. Invalid JSON. Original Error: ${err}`
      );
    }
  }
  return {};
}

/**
 * Parse a single response to the according response object type.
 * @param response The string representation of a single response.
 * @param entityToConstructorMap A map that holds the entity type to constructor mapping
 * @returns The parsed response in the according response object representation.
 */
export function parseResponse(
  response: string,
  entityToConstructorMap: MapType<Constructable<EntityV2>>
): WriteResponse | ReadResponse | ErrorResponse {
  const responseData = {
    body: parseResponseBody(response),
    httpCode: parseHttpCode(response)
  };

  // retrieve response
  if (responseData.httpCode === 200) {
    return {
      ...responseData,
      type: getConstructor(responseData.body, entityToConstructorMap),
      as: asReadResponse(responseData.body),
      isSuccess: () => true
    };
  }

  // change set sub response
  if (responseData.httpCode === 201 || responseData.httpCode === 204) {
    return {
      ...responseData,
      type: getConstructor(responseData.body, entityToConstructorMap),
      as: asWriteResponse(responseData.body)
    };
  }

  // error response
  return { ...responseData, isSuccess: () => false };
}

/**
 * Parse the complete batch HTTP response.
 * @param batchResponse HTTP response.
 * @param entityToConstructorMap A map that holds the entity type to constructor mapping.
 * @returns An array of parsed sub responses of the batch response.
 */
export function parseBatchResponse(
  batchResponse: HttpResponse,
  entityToConstructorMap: MapType<Constructable<EntityV2>>
): (ErrorResponse | ReadResponse | WriteResponses)[] {
  return splitBatchResponse(batchResponse).map(response => {
    const contentType = getHeaderValue('content-type', parseHeaders(response));

    // is change set response
    if (isMultipartContentType(contentType)) {
      return {
        responses: splitChangeSetResponse(response).map(
          subResponse =>
            parseResponse(subResponse, entityToConstructorMap) as WriteResponse
        ),
        isSuccess: () => true
      };
    }

    // is read or error response
    if (isHttpContentType(contentType)) {
      return parseResponse(response, entityToConstructorMap) as
        | ReadResponse
        | ErrorResponse;
    }

    throw Error(
      `Cannot parse batch response. Unknown subresponse 'Content-Type' header '${contentType}'.`
    );
  });
}

function isMultipartContentType(contentType: string): boolean {
  return contentType.trim().startsWith('multipart/mixed');
}

function isHttpContentType(contentType: string): boolean {
  return contentType.trim().startsWith('application/http');
}
