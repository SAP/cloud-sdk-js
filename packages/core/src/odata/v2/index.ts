/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

export * from './edm-types';
export * from './entity';
export * from './entity-deserializer';
export * from './entity-serializer';
export * from './filter-functions';
export * from './odata-batch-change-set';
export * from './odata-batch-request-builder';
export * from './odata-batch-retrieve-request';
export * from './payload-value-converter';
export * from './request-builder';
export * from './selectable/custom-field';
export * from './uri-conversion';
export * from './legacy-request-configs';
export * from './extract-odata-etag';
export * from '../common';
