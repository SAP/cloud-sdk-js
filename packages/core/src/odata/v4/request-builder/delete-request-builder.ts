/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import { errorWithCause } from '@sap-cloud-sdk/util';
import { Constructable, EntityIdentifiable, FieldType } from '../../common';
import { MethodRequestBuilderBase } from '../../common/request-builder/request-builder-base';
import { ODataDeleteRequestConfig } from '../../common/request/odata-delete-request-config';
import { EntityV4 } from '../entity';
import { DestinationOptions } from '../../../scp-cf';
import {
  Destination,
  DestinationNameAndJwt
} from '../../../scp-cf/destination-service-types';
import { oDataUriV4 } from '../uri-conversion';
/**
 * Create OData query to delete an entity.
 *
 * @typeparam EntityT - Type of the entity to be deleted
 */
export class DeleteRequestBuilderV4<EntityT extends EntityV4>
  extends MethodRequestBuilderBase<ODataDeleteRequestConfig<EntityT>>
  implements EntityIdentifiable<EntityT> {
  readonly _entityConstructor: Constructable<EntityT>;
  readonly _entity: EntityT;

  /**
   * Creates an instance of DeleteRequestBuilder. If the entity is passed, version identifier will also be added.
   *
   * @param entityConstructor - Constructor type of the entity to be deleted
   * @param keysOrEntity - Entity or Key-value pairs of key properties for the given entity
   */
  constructor(
    entityConstructor: Constructable<EntityT>,
    keysOrEntity: Record<string, FieldType> | EntityV4
  ) {
    super(new ODataDeleteRequestConfig(entityConstructor, oDataUriV4));
    this._entityConstructor = entityConstructor;

    if (keysOrEntity instanceof EntityV4) {
      this.requestConfig.keys = oDataUriV4.getEntityKeys(
        keysOrEntity,
        entityConstructor
      );
      this.setVersionIdentifier(keysOrEntity.versionIdentifier);
    } else {
      this.requestConfig.keys = keysOrEntity;
    }
  }

  /**
   * Add ETag version identifier in the delete request header.
   *
   * @param etag - The version identifier of the entity
   * @returns The builder itself, to facilitate method chaining
   */
  setVersionIdentifier(etag: string): this {
    if (etag) {
      this.requestConfig.addCustomHeaders({ 'if-match': etag });
    }
    return this;
  }

  /**
   * Instructs the request to force an overwrite of the entity by sending an 'If-Match: *' header instead of sending the ETag version identifier.
   *
   * @returns this The request itself to ease chaining while executing the request
   */
  ignoreVersionIdentifier(): this {
    this.requestConfig.versionIdentifierIgnored = true;
    return this;
  }

  /**
   * Execute query.
   *
   * @param destination - Destination to execute the request against
   * @param options - Options to employ when fetching destinations
   * @returns A promise resolving once the entity was deleted
   */
  async execute(
    destination: Destination | DestinationNameAndJwt,
    options?: DestinationOptions
  ): Promise<void> {
    return this.build(destination, options)
      .then(request => request.execute())
      .then(() => Promise.resolve())
      .catch(error =>
        Promise.reject(errorWithCause('OData delete request failed!', error))
      );
  }
}
