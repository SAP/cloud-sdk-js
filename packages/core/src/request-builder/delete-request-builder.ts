/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import { errorWithCause, MapType } from '@sap-cloud-sdk/util';
import { Constructable } from '../constructable';
import { Entity, EntityIdentifiable } from '../entity';
import { DestinationOptions } from '../scp-cf';
import { Destination, DestinationNameAndJwt } from '../scp-cf/destination-service-types';
import { FieldType } from '../selectable/field';
import { MethodRequestBuilderBase } from './request-builder-base';
import { getEntityKeys } from './request/get-keys';
import { ODataDeleteRequestConfig } from './request/odata-delete-request-config';

/**
 * Create OData query to delete an entity.
 *
 * @typeparam EntityT - Type of the entity to be deleted
 */
export class DeleteRequestBuilder<EntityT extends Entity> extends MethodRequestBuilderBase<ODataDeleteRequestConfig<EntityT>>
  implements EntityIdentifiable<EntityT> {
  readonly _entityConstructor: Constructable<EntityT>;

  /**
   * Creates an instance of DeleteRequestBuilder. If the entity is passed, version identifier will also be added.
   *
   * @param entityConstructor - Constructor type of the entity to be deleted
   * @param keysOrEntity - Entity or Key-value pairs of key properties for the given entity
   */
  constructor(entityConstructor: Constructable<EntityT>, keysOrEntity: MapType<FieldType> | Entity) {
    super(new ODataDeleteRequestConfig(entityConstructor));
    this._entityConstructor = entityConstructor;

    if (keysOrEntity instanceof Entity) {
      this.requestConfig.keys = getEntityKeys(keysOrEntity, entityConstructor);
      this.setVersionIdentifier(keysOrEntity.versionIdentifier);
    } else {
      this.requestConfig.keys = keysOrEntity;
    }
  }

  /**
   * Add ETag version identifier in the delete request header.
   *
   * @param etag - The version identifier of the entity
   * @returns The builder itself, to facilitate method chaining
   */
  setVersionIdentifier(etag: string): this {
    if (etag) {
      this.requestConfig.addCustomHeaders({ 'if-match': etag });
    }
    return this;
  }

  /**
   * Instructs the request to force an overwrite of the entity by sending an 'If-Match: *' header instead of sending the ETag version identifier.
   *
   * @returns this The request itself to ease chaining while executing the request
   */
  ignoreVersionIdentifier(): this {
    this.requestConfig.versionIdentifierIgnored = true;
    return this;
  }

  /**
   * Execute query.
   *
   * @param destination - Destination to execute the request against
   * @param options - Options to employ when fetching destinations
   * @returns A promise resolving once the entity was deleted
   */
  async execute(destination: Destination | DestinationNameAndJwt, options?: DestinationOptions): Promise<void> {
    return this.build(destination, options)
      .then(request => request.execute())
      .then(() => Promise.resolve())
      .catch(error => Promise.reject(errorWithCause('OData delete request failed!', error)));
  }
}
