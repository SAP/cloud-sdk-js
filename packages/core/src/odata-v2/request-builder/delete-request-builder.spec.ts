import nock from 'nock';
import { v4 as uuid } from 'uuid';
import {
  defaultDestination,
  mockDeleteRequest
} from '../../../test/test-util/request-mocker';
import { testEntityResourcePath } from '../../../test/test-util/test-data';
import { TestEntity } from '../../../test/test-util/test-services/v2/test-service';
import { DeleteRequestBuilderV2 } from './delete-request-builder';

describe('DeleteRequestBuilderV2', () => {
  const keyPropGuid = uuid();
  const keyPropString = 'TEST_ID';

  afterAll(() => {
    nock.cleanAll();
  });

  it('delete request with keys should resolve', async () => {
    mockDeleteRequest({
      path: testEntityResourcePath(keyPropGuid, keyPropString)
    });

    const deleteRequest = new DeleteRequestBuilderV2(TestEntity, {
      KeyPropertyGuid: keyPropGuid,
      KeyPropertyString: keyPropString
    }).execute(defaultDestination);

    await expect(deleteRequest).resolves.toBe(undefined);
  });

  it('delete request with entity and version identifier should resolve', async () => {
    const versionId = 'not-a-star';
    const entity = TestEntity.builder()
      .keyPropertyGuid(keyPropGuid)
      .keyPropertyString(keyPropString)
      .build()
      .setVersionIdentifier(versionId);

    mockDeleteRequest({
      path: testEntityResourcePath(keyPropGuid, keyPropString),
      additionalHeaders: {
        'if-match': versionId
      }
    });

    const deleteRequest = new DeleteRequestBuilderV2(
      TestEntity,
      entity
    ).execute(defaultDestination);

    await expect(deleteRequest).resolves.toBe(undefined);
  });

  it('delete request with version identifier on the request should resolve', async () => {
    const versionId = 'not-a-star';

    mockDeleteRequest({
      path: testEntityResourcePath(keyPropGuid, keyPropString),
      additionalHeaders: {
        'if-match': versionId
      }
    });

    const deleteRequest = new DeleteRequestBuilderV2(TestEntity, {
      KeyPropertyGuid: keyPropGuid,
      KeyPropertyString: keyPropString
    })
      .setVersionIdentifier('not-a-star')
      .execute(defaultDestination);

    await expect(deleteRequest).resolves.toBe(undefined);
  });

  it('delete requests does not use if-match header when the version identifier is an empty string', async () => {
    mockDeleteRequest({
      path: testEntityResourcePath(keyPropGuid, keyPropString)
    });

    const deleteRequest = new DeleteRequestBuilderV2(TestEntity, {
      KeyPropertyGuid: keyPropGuid,
      KeyPropertyString: keyPropString
    })
      .setVersionIdentifier('')
      .execute(defaultDestination);

    await expect(deleteRequest).resolves.toBe(undefined);
  });

  it('should ignore the version identifier on delete if set', async () => {
    mockDeleteRequest({
      path: testEntityResourcePath(keyPropGuid, keyPropString),
      additionalHeaders: {
        'if-match': '*'
      }
    });

    const deleteRequest = new DeleteRequestBuilderV2(TestEntity, {
      KeyPropertyGuid: keyPropGuid,
      KeyPropertyString: keyPropString
    })
      .ignoreVersionIdentifier()
      .execute(defaultDestination);

    await expect(deleteRequest).resolves.toBe(undefined);
  });

  it('throws an error when request execution fails', async () => {
    mockDeleteRequest({
      path: testEntityResourcePath(keyPropGuid, keyPropString),
      statusCode: 500
    });

    const deleteRequest = new DeleteRequestBuilderV2(TestEntity, {
      KeyPropertyGuid: keyPropGuid
    }).execute(defaultDestination);

    await expect(deleteRequest).rejects.toThrowErrorMatchingSnapshot();
  });
});
