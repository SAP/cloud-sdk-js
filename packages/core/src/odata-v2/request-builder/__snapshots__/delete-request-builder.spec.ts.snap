// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteRequestBuilder throws an error when request execution fails 1`] = `"OData delete request failed!"`;
